// @wokwi/elements/dist/cjs/7segment-element.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const lit_element_1 = require("lit-element");
let SevenSegmentElement = class SevenSegmentElement extends lit_element_1.LitElement {
    constructor() {
        super(...arguments);
        this.color = 'red';
        this.values = [0, 0, 0, 0, 0, 0, 0, 0];
    }
    static get styles() {
        return lit_element_1.css `
      polygon {
        transform: scale(0.9);
        transform-origin: 50% 50%;
        transform-box: fill-box;
      }
    `;
    }
    render() {
        const { color, values } = this;
        const fill = (index) => (values[index] ? color : '#ddd');
        return lit_element_1.html `
      <svg
        width="12mm"
        height="18.5mm"
        version="1.1"
        viewBox="0 0 12 18.5"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g transform="skewX(-8) translate(2, 0)">
          <polygon points="2 0 8 0 9 1 8 2 2 2 1 1" fill="${fill(0)}" />
          <polygon points="10 2 10 8 9 9 8 8 8 2 9 1" fill="${fill(1)}" />
          <polygon points="10 10 10 16 9 17 8 16 8 10 9 9" fill="${fill(2)}" />
          <polygon points="8 18 2 18 1 17 2 16 8 16 9 17" fill="${fill(3)}" />
          <polygon points="0 16 0 10 1 9 2 10 2 16 1 17" fill="${fill(4)}" />
          <polygon points="0 8 0 2 1 1 2 2 2 8 1 9" fill=${fill(5)} />
          <polygon points="2 8 8 8 9 9 8 10 2 10 1 9" fill=${fill(6)} />
        </g>
        <circle cx="11" cy="17" r="1.1" fill="${fill(7)}" />
      </svg>
    `;
    }
};
__decorate([
    lit_element_1.property()
], SevenSegmentElement.prototype, "color", void 0);
__decorate([
    lit_element_1.property({ type: Array })
], SevenSegmentElement.prototype, "values", void 0);
SevenSegmentElement = __decorate([
    lit_element_1.customElement('wokwi-7segment')
], SevenSegmentElement);
exports.SevenSegmentElement = SevenSegmentElement;
